core/lib/src/shutdown.rs

```diff
@@ -0,0 +1,92 @@
+use futures::{channel::mpsc, stream::StreamExt};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Wrapper around an mpsc channel.
+#[derive(Debug)]
+crate struct Shutdown {
+    sender: mpsc::Sender<()>,
+    crate receiver: Option<Receiver>,
+}
+
+impl Shutdown {
+    /// Create a `Shutdown`.
+    #[inline]
+    crate fn new() -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            sender,
+            receiver: Some(Receiver { receiver }),
+        }
+    }
+
+    /// Get a handle that can be used to call `.shutdown()`.
+    #[inline]
+    crate fn get_handle(&self) -> ShutdownHandle<'_> {
+        ShutdownHandle {
+            sender: &self.sender,
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # #![feature(proc_macro_hygiene, async_await)]
+/// # #[macro_use] extern crate rocket;
+/// #
+/// use rocket::shutdown::ShutdownHandle;
+///
+/// #[get("/shutdown")]
+/// fn shutdown(handle: &ShutdownHandle) -> &'static str {
+///     handle.shutdown();
+///     "Shutting down..."
+/// }
+///
+/// fn main() {
+///     # if false {
+///     rocket::ignite()
+///         .mount("/", routes![shutdown])
+///         .launch();
+///     # }
+/// }
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ShutdownHandle<'a> {
+    sender: &'a mpsc::Sender<()>,
+}
+
+impl ShutdownHandle<'_> {
+    /// Notify Rocket to shut down gracefully.
+    ///
+    /// This method should only need to be called once.
+    #[inline]
+    pub fn shutdown(&self) {
+        // Clone to avoid requiring `self` to be mutable.
+        self.sender
+            .clone()
+            .try_send(())
+            .expect("Error sending shutdown signal")
```

Comment:
I think we should purposefully ignore this error - with a good comment explaining why. An error means that either the send queue is full (too many outstanding shutdown requests) or the receiver was dropped (already shut down).